sa/model.go

```diff
@@ -396,3 +400,152 @@ func modelToOrder(om *orderModel) (*corepb.Order, error) {
+	0: "http-01",
+	1: "tls-sni-01",
+	2: "dns-01",
+	3: "tls-alpn-01",
+}
+
+type authz2Model struct {
+	ID               int64
+	IdentifierType   string
+	IdentifierValue  string
+	RegistrationID   int64
+	Status           string
+	Expires          *time.Time
+	Challenges       byte
+	Attempted        string
+	Token            []byte
+	ValidationError  []byte
+	ValidationRecord []byte
+}
+
+func authzPBToModel(authz *corepb.Authorization) (*authz2Model, error) {
+	if authz.V2 == nil || !*authz.V2 {
+		return nil, errors.New("authorization is not v2 format")
+	}
+	expires := time.Unix(0, *authz.Expires)
+	id, err := strconv.Atoi(*authz.Id)
+	if err != nil {
+		return nil, err
+	}
+	am := &authz2Model{
+		ID:              int64(id),
+		IdentifierValue: *authz.Identifier,
+		RegistrationID:  *authz.RegistrationID,
+		Status:          *authz.Status,
+		Expires:         &expires,
+	}
+	var tokenStr string
+	for _, chall := range authz.Challenges {
+		am.Challenges |= 1 << challTypeToBit[*chall.Type]
+		tokenStr = *chall.Token
+		if *chall.Status == string(core.StatusValid) || *chall.Status == string(core.StatusInvalid) {
+			if am.Attempted != "" || am.ValidationRecord != nil || am.ValidationError != nil {
+				return nil, errors.New("multiple challenges are non-pending")
+			}
+			am.Attempted = *chall.Type
+			records := make([]core.ValidationRecord, len(chall.Validationrecords))
+			for i, recordPB := range chall.Validationrecords {
+				var err error
+				records[i], err = grpc.PBToValidationRecord(recordPB)
+				if err != nil {
+					return nil, err
+				}
+			}
+			var err error
+			am.ValidationRecord, err = json.Marshal(records)
+			if err != nil {
+				return nil, err
+			}
+			if chall.Error != nil {
+				prob, err := grpc.PBToProblemDetails(chall.Error)
+				if err != nil {
+					return nil, err
+				}
+				am.ValidationError, err = json.Marshal(prob)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+		token, err := base64.StdEncoding.DecodeString(tokenStr)
+		if err != nil {
+			return nil, err
+		}
+		am.Token = token
+	}
+
+	return am, nil
+}
+
+func modelToAuthzPB(am *authz2Model) (*corepb.Authorization, error) {
+	expires := am.Expires.UnixNano()
+	id := fmt.Sprintf("%d", am.ID)
+	v2 := true
+	pb := &corepb.Authorization{
+		V2:             &v2,
+		Id:             &id,
+		Status:         &am.Status,
+		Identifier:     &am.IdentifierValue,
+		RegistrationID: &am.RegistrationID,
+		Expires:        &expires,
+	}
+	challenges := int64(0)
+	for pos := uint(0); pos < 8; pos++ {
+		if (am.Challenges>>pos)&1 == 1 {
+			challType := bitToChallType[pos]
+			status := string(core.StatusPending)
+			id := challenges
+			challenges++
+			token := base64.StdEncoding.EncodeToString(am.Token)
+			challenge := &corepb.Challenge{
+				Id:     &id,
+				Type:   &challType,
+				Status: &status,
+				Token:  &token,
+			}
+			if am.Attempted != "" && am.Attempted == challType {
+				if len(am.ValidationError) != 0 {
```

Comment:
Will this `if/else` pair correctly handle a previously valid authorization that was deactivated?

Here's my thinking: If the authz was valid it would have no ValidationError and a non-empty "Attempted" value from the challenge type used. After the authz was deactivated it would still have a non-empty attempted value and an empty validation error but the `else` clause here would set it to `core.StatusValid` no?

Reply:
This is for the challenge status though, which should still be valid even if the authorization is deactivated I believe.

Reply:
Ahhhh, good point. I overlooked this was challenge status :+1: